File: src/系统/11．剧情系统/00．公共/04．剧情单位移动控制.lua
local ____lualib = require("lualib_bundle")
local __TS__ArrayIndexOf = ____lualib.__TS__ArrayIndexOf
local __TS__ArraySplice = ____lualib.__TS__ArraySplice
local ____exports = {}
---
-- @noSelfInFile
local jass = require("jass.common")
local ____require_result_0 = require("系统.00．核心系统.05．中心计时器")
local addPeriodicCallback = ____require_result_0.addPeriodicCallback
local removePeriodicCallback = ____require_result_0.removePeriodicCallback
local GetUnitTypeId = jass.GetUnitTypeId
local GetUnitState = jass.GetUnitState
local GetUnitX = jass.GetUnitX
local GetUnitY = jass.GetUnitY
local IssueImmediateOrder = jass.IssueImmediateOrder
local IssuePointOrder = jass.IssuePointOrder
local SetUnitFacing = jass.SetUnitFacing
local UNIT_STATE_LIFE = jass.UNIT_STATE_LIFE
local _____8FD0_884C_4E2D_79FB_52A8_4EFB_52A1_5217_8868 = {}
local function _____5355_4F4D_53EF_79FB_52A8(unit)
    return unit ~= nil and unit ~= 0 and GetUnitTypeId(unit) ~= 0 and GetUnitState(unit, UNIT_STATE_LIFE) > 0.405
end
local function _____4ECE_8FD0_884C_5217_8868_79FB_9664(_____4EFB_52A1)
    local index = __TS__ArrayIndexOf(_____8FD0_884C_4E2D_79FB_52A8_4EFB_52A1_5217_8868, _____4EFB_52A1)
    if index >= 0 then
        __TS__ArraySplice(_____8FD0_884C_4E2D_79FB_52A8_4EFB_52A1_5217_8868, index, 1)
    end
end
local function _____505C_6B62_79FB_52A8_4EFB_52A1(_____4EFB_52A1, _____6267_884C_4E2D_6B62_56DE_8C03)
    if not _____4EFB_52A1["运行中"] then
        return
    end
    _____4EFB_52A1["运行中"] = false
    if _____4EFB_52A1["回调ID"] ~= 0 then
        removePeriodicCallback(_____4EFB_52A1["回调ID"])
        _____4EFB_52A1["回调ID"] = 0
    end
    _____4ECE_8FD0_884C_5217_8868_79FB_9664(_____4EFB_52A1)
    if _____6267_884C_4E2D_6B62_56DE_8C03 and _____4EFB_52A1["配置"]["中止回调"] ~= nil then
        _____4EFB_52A1["配置"]["中止回调"](_____4EFB_52A1["配置"]["单位"])
    end
end
local function ____on_5267_60C5_5355_4F4D_79FB_52A8Tick(variable)
    local _____4EFB_52A1 = variable
    if _____4EFB_52A1 == nil or not _____4EFB_52A1["运行中"] then
        return
    end
    local _____914D_7F6E = _____4EFB_52A1["配置"]
    if not _____5355_4F4D_53EF_79FB_52A8(_____914D_7F6E["单位"]) or _____914D_7F6E["是否继续"] ~= nil and not _____914D_7F6E["是否继续"](_____914D_7F6E["单位"]) then
        _____505C_6B62_79FB_52A8_4EFB_52A1(_____4EFB_52A1, true)
        return
    end
    local dx = GetUnitX(_____914D_7F6E["单位"]) - _____914D_7F6E["目标X"]
    local dy = GetUnitY(_____914D_7F6E["单位"]) - _____914D_7F6E["目标Y"]
    local _____5230_8FBE_8DDD_79BB = _____914D_7F6E["到达距离"] or 96
    if dx * dx + dy * dy <= _____5230_8FBE_8DDD_79BB * _____5230_8FBE_8DDD_79BB then
        _____4EFB_52A1["已到达"] = true
        local _____5230_8FBE_547D_4EE4 = _____914D_7F6E["到达命令"] == nil and "holdposition" or _____914D_7F6E["到达命令"]
        if _____5230_8FBE_547D_4EE4 ~= false then
            IssueImmediateOrder(_____914D_7F6E["单位"], _____5230_8FBE_547D_4EE4)
        end
        if _____914D_7F6E["到达朝向"] ~= nil then
            SetUnitFacing(_____914D_7F6E["单位"], _____914D_7F6E["到达朝向"])
        end
        _____505C_6B62_79FB_52A8_4EFB_52A1(_____4EFB_52A1, false)
        if _____914D_7F6E["到达回调"] ~= nil then
            _____914D_7F6E["到达回调"](_____914D_7F6E["单位"])
        end
        return
    end
    if _____914D_7F6E["补发移动命令"] ~= false then
        IssuePointOrder(_____914D_7F6E["单位"], _____914D_7F6E["移动命令"] or "move", _____914D_7F6E["目标X"], _____914D_7F6E["目标Y"])
    end
end
--- 驱动剧情单位走到目标并可靠停留。同一单位开始新任务时会自动取消旧任务，
-- 所有完成、失效和手动取消路径都会注销中心计时器回调。
____exports["开始剧情单位保持移动"] = function(_____914D_7F6E)
    if not _____5355_4F4D_53EF_79FB_52A8(_____914D_7F6E["单位"]) then
        return nil
    end
    do
        local i = #_____8FD0_884C_4E2D_79FB_52A8_4EFB_52A1_5217_8868 - 1
        while i >= 0 do
            local _____65E7_4EFB_52A1 = _____8FD0_884C_4E2D_79FB_52A8_4EFB_52A1_5217_8868[i + 1]
            if _____65E7_4EFB_52A1["配置"]["单位"] == _____914D_7F6E["单位"] then
                _____505C_6B62_79FB_52A8_4EFB_52A1(_____65E7_4EFB_52A1, false)
            end
            i = i - 1
        end
    end
    local _____4EFB_52A1 = {}
    local _____63A7_5236_5668 = {
        ["取消"] = function() return _____505C_6B62_79FB_52A8_4EFB_52A1(_____4EFB_52A1, false) end,
        ["是否运行"] = function() return _____4EFB_52A1["运行中"] end,
        ["是否到达"] = function() return _____4EFB_52A1["已到达"] end
    }
    _____4EFB_52A1["配置"] = _____914D_7F6E
    _____4EFB_52A1["控制器"] = _____63A7_5236_5668
    _____4EFB_52A1["回调ID"] = 0
    _____4EFB_52A1["运行中"] = true
    _____4EFB_52A1["已到达"] = false
    _____8FD0_884C_4E2D_79FB_52A8_4EFB_52A1_5217_8868[#_____8FD0_884C_4E2D_79FB_52A8_4EFB_52A1_5217_8868 + 1] = _____4EFB_52A1
    IssuePointOrder(_____914D_7F6E["单位"], _____914D_7F6E["移动命令"] or "move", _____914D_7F6E["目标X"], _____914D_7F6E["目标Y"])
    _____4EFB_52A1["回调ID"] = addPeriodicCallback(_____914D_7F6E["检查间隔毫秒"] or 400, ____on_5267_60C5_5355_4F4D_79FB_52A8Tick, _____4EFB_52A1)
    return _____63A7_5236_5668
end
return ____exports

File: TS/系统/11．剧情系统/00．公共/04．剧情单位移动控制.ts
/** @noSelfInFile */

const jass = require("jass.common") as any;
const { addPeriodicCallback, removePeriodicCallback } = require("系统.00．核心系统.05．中心计时器") as {
  addPeriodicCallback: (this: void, intervalMs: number, callback: (this: void, variable?: any) => void, variable?: any) => number;
  removePeriodicCallback: (this: void, id: number) => void;
};

const GetUnitTypeId = jass.GetUnitTypeId as (this: void, unit: any) => number;
const GetUnitState = jass.GetUnitState as (this: void, unit: any, state: any) => number;
const GetUnitX = jass.GetUnitX as (this: void, unit: any) => number;
const GetUnitY = jass.GetUnitY as (this: void, unit: any) => number;
const IssueImmediateOrder = jass.IssueImmediateOrder as (this: void, unit: any, order: string) => boolean;
const IssuePointOrder = jass.IssuePointOrder as (this: void, unit: any, order: string, x: number, y: number) => boolean;
const SetUnitFacing = jass.SetUnitFacing as (this: void, unit: any, facing: number) => void;
const UNIT_STATE_LIFE = jass.UNIT_STATE_LIFE as any;

export interface 剧情单位移动配置 {
  单位: any;
  目标X: number;
  目标Y: number;
  到达距离?: number;
  检查间隔毫秒?: number;
  移动命令?: string;
  /** 默认持续补发移动命令，避免中立单位的原生游荡逻辑重新接管。 */
  补发移动命令?: boolean;
  /** 默认抵达后发布 holdposition；传入 false 可关闭。 */
  到达命令?: string | false;
  到达朝向?: number;
  /** 返回 false 时立即中止并排泄周期任务。 */
  是否继续?: (this: void, unit: any) => boolean;
  到达回调?: (this: void, unit: any) => void;
  中止回调?: (this: void, unit: any) => void;
}

export interface 剧情单位移动控制器 {
  取消(this: void): void;
  是否运行(this: void): boolean;
  是否到达(this: void): boolean;
}

interface 运行中移动任务 {
  配置: 剧情单位移动配置;
  控制器: 剧情单位移动控制器;
  回调ID: number;
  运行中: boolean;
  已到达: boolean;
}

const 运行中移动任务列表: 运行中移动任务[] = [];

function 单位可移动(this: void, unit: any): boolean {
  return unit != null && unit !== 0 && GetUnitTypeId(unit) !== 0 && GetUnitState(unit, UNIT_STATE_LIFE) > 0.405;
}

function 从运行列表移除(this: void, 任务: 运行中移动任务): void {
  const index = 运行中移动任务列表.indexOf(任务);
  if (index >= 0) 运行中移动任务列表.splice(index, 1);
}

function 停止移动任务(this: void, 任务: 运行中移动任务, 执行中止回调: boolean): void {
  if (!任务.运行中) return;
  任务.运行中 = false;
  if (任务.回调ID !== 0) {
    removePeriodicCallback(任务.回调ID);
    任务.回调ID = 0;
  }
  从运行列表移除(任务);
  if (执行中止回调 && 任务.配置.中止回调 != null) 任务.配置.中止回调(任务.配置.单位);
}

function on剧情单位移动Tick(this: void, variable?: any): void {
  const 任务 = variable as 运行中移动任务 | undefined;
  if (任务 == null || !任务.运行中) return;

  const 配置 = 任务.配置;
  if (!单位可移动(配置.单位) || (配置.是否继续 != null && !配置.是否继续(配置.单位))) {
    停止移动任务(任务, true);
    return;
  }

  const dx = GetUnitX(配置.单位) - 配置.目标X;
  const dy = GetUnitY(配置.单位) - 配置.目标Y;
  const 到达距离 = 配置.到达距离 ?? 96;
  if (dx * dx + dy * dy <= 到达距离 * 到达距离) {
    任务.已到达 = true;
    const 到达命令 = 配置.到达命令 === undefined ? "holdposition" : 配置.到达命令;
    if (到达命令 !== false) IssueImmediateOrder(配置.单位, 到达命令);
    if (配置.到达朝向 != null) SetUnitFacing(配置.单位, 配置.到达朝向);
    停止移动任务(任务, false);
    if (配置.到达回调 != null) 配置.到达回调(配置.单位);
    return;
  }

  if (配置.补发移动命令 !== false) {
    IssuePointOrder(配置.单位, 配置.移动命令 ?? "move", 配置.目标X, 配置.目标Y);
  }
}

/**
 * 驱动剧情单位走到目标并可靠停留。同一单位开始新任务时会自动取消旧任务，
 * 所有完成、失效和手动取消路径都会注销中心计时器回调。
 */
export function 开始剧情单位保持移动(this: void, 配置: 剧情单位移动配置): 剧情单位移动控制器 | undefined {
  if (!单位可移动(配置.单位)) return undefined;

  for (let i = 运行中移动任务列表.length - 1; i >= 0; i--) {
    const 旧任务 = 运行中移动任务列表[i];
    if (旧任务.配置.单位 === 配置.单位) 停止移动任务(旧任务, false);
  }

  const 任务 = {} as 运行中移动任务;
  const 控制器: 剧情单位移动控制器 = {
    取消: () => 停止移动任务(任务, false),
    是否运行: () => 任务.运行中,
    是否到达: () => 任务.已到达,
  };
  任务.配置 = 配置;
  任务.控制器 = 控制器;
  任务.回调ID = 0;
  任务.运行中 = true;
  任务.已到达 = false;
  运行中移动任务列表.push(任务);

  IssuePointOrder(配置.单位, 配置.移动命令 ?? "move", 配置.目标X, 配置.目标Y);
  任务.回调ID = addPeriodicCallback(配置.检查间隔毫秒 ?? 400, on剧情单位移动Tick, 任务);
  return 控制器;
}

export {};
